src/mqttproto/client_state_machine.py

```diff
@@ -151,6 +157,7 @@ def publish(
 
         """
         self._out_require_state(MQTTClientState.CONNECTED)
+        qos = min(qos, self._maximum_qos)
```

Comment:
I'm really not sure that silently downgrading the QoS is a good idea. How do other MQTT clients deal with this?

Reply:
I looked at paho-mqtt and it doesn't look like they're doing this. Doing so in the state machine is even worse than in the client implementation. Exposing the `maximum_qos` information was a good call, however.

Reply:
What the state machine could do is raise an error when attempting to send a message with an unsupported QoS.

Reply:
Or maybe we should just drop this line and let the client app deal with the fallout instead, same as with the Retain flag handling.

Reply:
Maybe, yeah, so long as we expose the maximum support QoS, this would seem reasonable.